### PR TITLE
FOGL-2614 browser system API tests added 

### PR DESCRIPTION
**Pull request by ashish-jabble**
**Test using sqlite (default):**
pytest test_browser_assets.py

**Test using postgres :**
pytest test_browser_assets.py --storage-plugin=postgres
 
**Results**

```
collected 42 items                                                                                                                           

../TestBrowserAssets test_get_asset_counts <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_query_bad_data <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_reading <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_query <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_summary <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_summary_invalid_sensor <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_readings_summary <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_time_limit <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_group_sec <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_group_min <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_series_query_group_hrs <- api/test_browser_assets.py PASSED
../TestBrowserAssets test_get_asset_sensor_readings_invalid_group <- api/test_browser_assets.py PASSED

== 42 passed in 32.19 seconds ==
```

### CODE REVIEW

File: tests/system/python/api/test_browser_assets.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert len(jdoc), "No data found"
+        assert SENSOR_VALUE == jdoc[0]['average']
+        assert SENSOR_VALUE == jdoc[0]['max']
+        assert SENSOR_VALUE == jdoc[0]['min']
+
+    # TODO: Add more paramterized tests with additional params like limit, skip, group, timeunits etc.
+    # And supply more data (nested readings) instead of 1 single -level reading
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: yes, let's supply atleast 3 values e.g. 1, 2, 0 and verify avg 1, max 2 and min 0. if we can use int and float both, that will be more valuable.

File: tests/system/python/api/test_browser_assets.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+
+__author__ = "Ashish Jabble"
+__copyright__ = "Copyright (c) 2019 Dianomic Systems"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+TEMPLATE_NAME = "template.json"
+SENSOR_VALUE = 40
+ASSET_NAME = 'Browser'
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: better to have asset name with space or much better to have  `sensor #1`

File: tests/system/python/api/test_browser_assets.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        time.sleep(wait_time)
+
+        yield self.start_south
+
+        # Cleanup code that runs after the caller test is over
+        remove_data_file(fogbench_template_path)
+        remove_directories("/tmp/foglamp-south-{}".format(south_plugin))
+
+    def test_get_asset_counts(self, start_south, foglamp_url):
+        conn = http.client.HTTPConnection(foglamp_url)
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: can not we reuse this `conn`?

**Comment by Singhal-Vaibhav** [comment]
Currently we dont have a plugin in which we can manipulate user_ts. Its better to add a programmatic south services which can send data to an earlier timestamp and we can verify data grouping and summary based on hours/min/sec in a more logical way. CoAP plugin is not apt for that.

**Comment by MarkRiddoch** [comment]
> Currently we dont have a plugin in which we can manipulate user_ts. Its better to add a programmatic south services which can send data to an earlier timestamp and we can verify data grouping and summary based on hours/min/sec in a more logical way. CoAP plugin is not apt for that.
> cc @MarkRiddoch

I believe we can use the HTTP south to do this as it has the user_ts in the payload - we preserve this I believe.

File: tests/system/python/api/test_browser_assets.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        assert avg == summary['average']
+        assert max(SENSOR_VALUES) == summary['max']
+        assert min(SENSOR_VALUES) == summary['min']
+
+    def test_get_asset_readings_summary_invalid_sensor(self, foglamp_url):
+        conn = http.client.HTTPConnection(foglamp_url)
+        invalid_sensor = "invalid"
+        conn.request("GET", '/foglamp/asset/{}/{}/summary'.format(ASSET_NAME, invalid_sensor))
+        r = conn.getresponse()
+        assert 200 == r.status
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: why this is a 200; should not it be 404?

File: tests/system/python/api/test_browser_assets.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+    def test_get_asset_series(self, foglamp_url):
+        conn = http.client.HTTPConnection(foglamp_url)
+        conn.request("GET", '/foglamp/asset/{}/{}/series'.format(ASSET_NAME, SENSOR))
+        r = conn.getresponse()
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        i = 0
+        for val in SENSOR_VALUES:
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: Could you please add more info / docstring here

File: tests/system/python/data/testplugin.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: we should not use `test` name as prefix here.

File: tests/system/python/data/testplugin.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+# FOGLAMP_BEGIN
+# See: http://foglamp.readthedocs.io/
+# FOGLAMP_END
+
+""" Module for Test async mode plugin """
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: is this correct?

File: tests/system/python/data/testplugin.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+import asyncio
+import copy
+import uuid
+import logging
+import datetime
+import async_ingest
+
+from foglamp.common import logger
+from foglamp.plugins.common import utils
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: unused?

File: tests/system/python/data/testplugin.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    _task = asyncio.ensure_future(_start_aiotest(handle), loop=loop)
+
+    def run():
+        global loop
+        loop.run_forever()
+
+    t = Thread(target=run)
+    t.start()
+
+async def _start_aiotest(handle):
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: add  a new line

File: tests/system/python/data/testplugin.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+    ts_lst = []
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: ts_lst = list()

File: tests/system/python/data/testplugin.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    ts_lst.append(str(datetime.datetime.now(datetime.timezone.utc).astimezone()))
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: May be 
`from datetime import datetime as dt, timezone as tz, timedelta`
or just 
`from datetime import datetime, timezone, timedelta`

File: tests/system/python/data/testplugin.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    ts_lst.append(str(datetime.datetime.now(datetime.timezone.utc).astimezone() - datetime.timedelta(seconds=3)))
+    ts_lst.append(str(datetime.datetime.now(datetime.timezone.utc).astimezone() - datetime.timedelta(minutes=5)))
+    ts_lst.append(str(datetime.datetime.now(datetime.timezone.utc).astimezone() - datetime.timedelta(minutes=6)))
+    ts_lst.append(str(datetime.datetime.now(datetime.timezone.utc).astimezone() - datetime.timedelta(hours=3)))
+    ts_lst.append(str(datetime.datetime.now(datetime.timezone.utc).astimezone() - datetime.timedelta(hours=5)))
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: Please add a docstring; why such test data

File: tests/system/python/data/testplugin.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+                'key': str(uuid.uuid4()),
+                'readings': {"loudness": i}
+            })
+            async_ingest.ingest_callback(c_callback, c_ingest_ref, data)
+            await asyncio.sleep(0.1)
+
+        except (Exception, RuntimeError) as ex:
+            _LOGGER.exception("TEST exception: {}".format(str(ex)))
+            raise exceptions.DataRetrievalError(ex)
+        i += 1
```
Comment (praveen-garg) [inline tests/system/python/data/testplugin.py]: except - else:

File: tests/system/python/api/test_browser_assets.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+SENSOR = 'loudness'
+SENSOR_VALUES = [1, 2, 3, 4, 5, 6]
+SOUTH_PLUGIN_NAME = 'testplugin'
+SERVICE_NAME = 'TestBrowserAPI'
+
+
+def validate(date_text, _format):
+    try:
+        datetime.datetime.strptime(date_text, _format)
+        return True
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: except-else?

File: tests/system/python/api/test_browser_assets.py [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+__version__ = "${VERSION}"
+
+ASSET_NAME = 'test-loudness'
+SENSOR = 'loudness'
+SENSOR_VALUES = [1, 2, 3, 4, 5, 6]
+SOUTH_PLUGIN_NAME = 'testplugin'
+SERVICE_NAME = 'TestBrowserAPI'
+
+
+def validate(date_text, _format):
```
Comment (praveen-garg) [inline tests/system/python/api/test_browser_assets.py]: validate ?

**Comment by praveen-garg** [comment]
And also please run the same using Postgres storage.

File: tests/system/python/api/test_browser_assets.py (hunk 4 above)
Comment (Singhal-Vaibhav) [reply tests/system/python/api/test_browser_assets.py]: Architect decision, please see the [discussion](https://scaledb.atlassian.net/browse/FOGL-1292?focusedCommentId=39059&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-39059)

**Pull request merged**